utils/generic.py

```diff
@@ -308,3 +308,184 @@ def annotation_type_in_vcf_info(t: Any) -> bool:
             isinstance(t, hl.tset) or
             isinstance(t, hl.tbool)
             )
+
+
+def phase_by_transmission(
+        proband_call: hl.expr.CallExpression,
+        father_call: hl.expr.CallExpression,
+        mother_call: hl.expr.CallExpression,
+        phased_genotype_field: str = 'PBT_GT'
+) -> hl.MatrixTable:
+    """
+
+    Phase a trio MatrixTable by trio-based allele transmission. Currently uses a hardcalls to phase and only phases when all 3 members of the trio are present and have a genotype.
+    All calls need to be from the same trio MatrixTable.
+
+    Typical usage:
+    ```
+        tm = hl.trio_matrix(mt, ped)
+        phased_tm = phase_by_transmission(tm.proband_entry.GT, tm.father_entry.GT, tm.mother_entry.GT)
+    ```
+
+
+    :param CallExpression proband_call: Proband call to use for phasing
+    :param CallExpression father_call: Father call to use for phasing
+    :param CallExpression mother_call: Mother call to use for phasing
+    :param Pedigree ped: Pedigree to use to get the trios for phasing
+    :param str phased_genotype_field: name of the new phased genotype field to add
+    :return:
+    """
+
+    def phase_trio_genotypes(proband_gt: hl.expr.CallExpression,
+                             father_gt: hl.expr.CallExpression,
+                             mother_gt: hl.expr.CallExpression,
+                             n_alleles: int) -> hl.expr.ArrayExpression:
+        """
+        Given a trio genotype and the number of alleles at the site, phases the genotypes based on allele transmission in the trio.
+
+        :param CallExpression proband_gt: proband genotype
+        :param CallExpression father_gt: father genotype
+        :param CallExpression mother_gt: mother genotype
+        :param int n_alleles: Number of alleles at the site
+        :return: Array of genotypes (proband, father, mother)
+        :rtype: ArrayExpression
+        """
+
+        def gt_to_vectors(gt: hl.expr.CallExpression, n_alleles: int) -> hl.expr.ArrayExpression:
+            """
+            Get the set of all allele-vectors in a genotype.
+            It is returned as an ordered array where the first vector corresponds to the first allele,
+            and the second vector (only present if het) the second allele.
+
+            :param CallExpression gt: genotype
+            :param int n_alleles: number of alleles at the site
+            :return:
+            """
+            return hl.cond(
+                gt.is_het(),
+                hl.array([
+                    allele_to_vector(gt, 0, n_alleles),
+                    allele_to_vector(gt, 1, n_alleles)
+                ]),
+                hl.array([allele_to_vector(gt, 0, n_alleles)])
+            )
+
+        def allele_to_vector(gt: hl.expr.CallExpression, allele_index: int, n_alleles: int) -> hl.expr.ArrayExpression:
```

Comment:
Wait, I guess it kind of is. But then I'm confused as to the role of `allele_index`

Reply:
Ok after a bunch more reading, I understand a bit better, but might be good to document why you're doing this (here, or a short comment at the `combinations` line)

Reply:
Yeah, so I could also use something like
```
hl.call(gt[0]).one_hot_alleles(alleles)
hl.call(gt[1]).one_hot_alleles(alleles)
```

Maybe clearer? Otherwise, I'll rename this `one_hot_allele` to make it clearer

Reply:
Ooo yeah, if it's equivalent, I think that is better